src/internal.rs

```diff
@@ -473,17 +479,128 @@ pub struct Rav1dFrameContext_lf {
+        }
+    }
+}
+
+impl PartialOrd for Rav1dTaskIndex {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        match (self, other) {
+            (Self::Task(x), Self::Task(y)) => x.partial_cmp(y),
+            (Self::TileTask(x), Self::TileTask(y)) => x.partial_cmp(y),
+            _ => None,
+        }
+    }
+}
+
+impl Add<usize> for Rav1dTaskIndex {
+    type Output = Self;
+
+    fn add(self, rhs: usize) -> Self::Output {
+        match self {
+            Self::Task(i) => Self::Task(i + rhs),
+            Self::TileTask(i) => Self::TileTask(i + rhs),
+            Self::Init => panic!("Cannot index from the init task"),
+        }
+    }
+}
+
+impl AddAssign<usize> for Rav1dTaskIndex {
+    fn add_assign(&mut self, rhs: usize) {
+        match self {
+            Self::Task(i) => *i += rhs,
+            Self::TileTask(i) => *i += rhs,
+            Self::Init => panic!("Cannot index from the init task"),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct Rav1dTasks {
+    tasks: Vec<Rav1dTask>,
+    tile_tasks_vec: Vec<Rav1dTask>,
+    init_task: Rav1dTask,
+    pub tile_tasks: [Option<Rav1dTaskIndex>; 2],
+    pub head: Option<Rav1dTaskIndex>,
+    pub tail: Option<Rav1dTaskIndex>,
+    // Points to the task directly before the cur pointer in the queue.
+    // This cur pointer is theoretical here, we actually keep track of the
+    // "prev_t" variable. This is needed to not loose the tasks in
+    // [head;cur-1] when picking one for execution.
+    pub cur_prev: Option<Rav1dTaskIndex>,
+}
+
+impl Rav1dTasks {
+    pub fn grow_tasks(&mut self, new_len: usize) {
+        if new_len > self.tasks.len() {
+            self.tasks.clear();
+            self.tasks.resize_with(new_len, Default::default);
+        }
+    }
+
+    pub fn grow_tile_tasks(&mut self, new_len: usize) {
+        if new_len > self.tile_tasks_vec.len() {
+            self.tile_tasks_vec.clear();
+            self.tile_tasks_vec.resize_with(new_len, Default::default);
+            self.tile_tasks[0] = Some(Rav1dTaskIndex::TileTask(0));
+        }
+    }
+
+    pub fn init_task(&mut self) -> &mut Rav1dTask {
+        &mut self.init_task
+    }
+}
+
+impl Index<Rav1dTaskIndex> for Rav1dTasks {
+    type Output = Rav1dTask;
+
+    fn index(&self, index: Rav1dTaskIndex) -> &Self::Output {
+        match index {
+            Rav1dTaskIndex::Task(index) => &self.tasks[index],
+            Rav1dTaskIndex::TileTask(index) => &self.tile_tasks_vec[index],
+            Rav1dTaskIndex::Init => &self.init_task,
+        }
+    }
+}
+
+impl IndexMut<Rav1dTaskIndex> for Rav1dTasks {
+    fn index_mut(&mut self, index: Rav1dTaskIndex) -> &mut Self::Output {
+        match index {
+            Rav1dTaskIndex::Task(index) => &mut self.tasks[index],
+            Rav1dTaskIndex::TileTask(index) => &mut self.tile_tasks_vec[index],
+            Rav1dTaskIndex::Init => &mut self.init_task,
         }
```

Comment:
Too complicated, we shouldn't be doing optimizations prematurely (especially since the compiler could probably do this already here).

Reply:
We can save this for later since performance might not be a concern here, but I do think there's a benefit in approaching these things with performance in mind.  The compiler does not do this optimization to make it branchless, and so will incur mispredictions.